Update the struct definitions to match version 3.14 of the LSP spec

ahoppen opened the pull request:
I went through the spec, comparing the struct definitions to what is defined in https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/ and changed the definitions accordingly to match the spec.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

benlangmuir commented on Sources/LanguageServerProtocol/Character+Codable.swift:
```diff
@@ -0,0 +1,26 @@
+//
+//===----------------------------------------------------------------------===//
+
+extension Character: Codable {
```
We shouldn't tie Swift's definition of a Character to serialization - for one thing, the definition of a Character changes over time because of changes to unicode.  It's also unlikely to ever match what other lsp implementations use to define a character.  The spec itself uses String.

ahoppen commented:
I did the following change to address the review feedback:
- I changed the response of `DefinitionRequest` and `ImplementationRequest` to parse `Location` as `[Location]`
- Similarly `CompletionRequest` parses `[CompletionItem]` as `CompletionList`
- I removed the extension on `Optional` since it’s no longer necessary with your changes
- I added tests for any custom `Codable` conformance

I also split a couple commits off this PR to keep it at a more manageable size. They now live in #206.

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test
